### ISSUE TITLE
The current use of etcd Leases in Kubernetes violates the invariants needed for watch to function correctly

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#110210]
We're currently (potentially) attaching the same etcd Lease to multiple objects based on the LeaseManager configuration:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/lease_manager.go#L34

However, when the lease expires in etcd, all objects that are attached to that lease are actually deleted in a single transaction:
https://github.com/etcd-io/etcd/blob/e02ac59cfc4e750f4b0a2a121e20db8e6775d5cd/server/lease/lessor.go#L338-L356

This basically means, that deletions of all of them  will have exactly the same ResourceVersion.

The above violates the invariant that we need in order to ensure that watch requests can be resumed at any point:
- let's say that a watch breaks (e.g. TCP connection gets closed or sth) in the middle of processing objects from a given transaction (which by definition all have the same RV)
- then the client will try to resume from the last receive RV
- but that would mean that the ones from that transaction that wasn't processed previously will get missed and not delivered to that client

Fortunately, the TTL function (that is what decides whether an etcd Lease will be attached to the object or not) is only configured for events:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/event/storage/storage.go#L42

so we don't seem to be exposed for the bug in any core controllers.

That said we are still exposed to a bug that even when used correctly, WATCH requests for events may miss some DELETE events.

/sig api-machinery
/priority important-soon
/kind bug

@liggitt @deads2k @lavalamp @stevekuznetsov @kubernetes/sig-api-machinery-bugs 
@mm4tt @mborsz

**Comment by lavalamp** [comment]
I like the bookmark idea! But we would need to be selective and only do that for collections that have batched changes.

**Comment by wojtek-t** [comment]
> I like the bookmark idea! But we would need to be selective and only do that for collections that have batched changes.

Which at this point means pretty much only events...

**Comment by logicalhan** [comment]
We should consider using k8s leases and bind those to etcd events for expiration instead of native etcd leases. Then we would just garbage collect etcd events for deletion when the k8s leases expired (presumably in the gc controller) and it would delete the events sequentially, fixing the watch issue.

**Comment by wojtek-t** [comment]
Techncally k8s leases don't expire now (I mean, they are not removed on expiration).
So in order to do that, we would have to implment this mechanism first (in somewhat careful manner, because in case of leader-election we shouldn't actually be immediately removing leases on expiration).

If we would have the above, the mechanism for deletion already exists - it's called OwnerReferences :)
But I'm slightly concerned about the performance and failure modes of it:
- events and high-throughput objects, especially in large clusters - so it's potentially very high QPS from a single client (GC)
- it can either starve other things (preferable) or be starved itself (which would be serious failure mode as it can result in constantly growing storage)


Still - this can be our best option to do that in backward-compatible way, but if we decide for it we need to be very careful about scale/performance-related consequences.

**Comment by logicalhan** [comment]
Yeah, I was alluding to owner references when I mentioned:

> presumably in the gc controller

But yeah, I generally agree that the performance is probably the biggest concern. However this path does allow us to get to protocol correctness, which is nice.

**Comment by mm4tt** [comment]
> Another approach is to emit bookmarks between batches of resources that can be bulk-edited and recommend customers to use only bookmarks for recovery. Bookmark would be guaranteed to be emitted between batches (a bookmark with special field explaining the purpose).

The problem here is that it might a backward incompatible change. For k8s watches served from etcd (skipping api-server's watch cache), the client controls whether the bookmarks should be emitted or not - https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L347, https://github.com/etcd-io/etcd/blob/main/api/etcdserverpb/rpc.proto#L763. I know from experiments that ignoring this setting and always sending bookmark was breaking some older K8s clients - they were failing to decode the bookmark watch event.

**Comment by liggitt** [comment]
> The problem here is that it might a backward incompatible change.

Right... emitting bookmarks could *only* be done when the request indicated bookmarks were acceptable to return

**Comment by BenTheElder** [comment]
feels like another example for https://github.com/kubernetes/test-infra/issues/32957
